### PR TITLE
Add cupy func generate

### PR DESCRIPTION
**Pull request by delta2323**
This is re-submit of #1018, which had been merged, but was reverted as #1998. I and @hisa0507-san made the following changes from #1018:

* Explicitly write that the code is partially borrow codes from corresponding NumPy's objects.
* Move part of document of `CClass` and `RClass` to `c_` and `r_`, respectively.

### CODE REVIEW

File: cupy/indexing/generate.py [hunk 1]
```diff
@@ -1,13 +1,167 @@
...
+    """Translates slice objects to concatenation along an axis.
+
+    For detailed documentation on usage, see `r_`.
+    This implementation is partially borrowed from NumPy's one
+
+    """
+
+    def _output_obj(self, newobj, tempobj, ndmin, trans1d):
+        k2 = ndmin - tempobj.ndim
+        if (trans1d < 0):
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Remove paren.

File: cupy/indexing/generate.py [hunk 2]
```diff
@@ -1,13 +1,167 @@
...
+        objs = []
+        scalars = []
+        arraytypes = []
+        scalartypes = []
+        if isinstance(key, six.string_types):
+            raise NotImplementedError
+        if not isinstance(key, tuple):
+            key = (key,)
+
+        for k in six.moves.range(len(key)):
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Why not use `enumerate(key)`?

File: cupy/indexing/generate.py [hunk 3]
```diff
@@ -1,13 +1,167 @@
...
+        return cupy.concatenate(tuple(objs), axis=self.axis)
+
+    def __len__(self):
+        return 0
+
+
+class CClass(AxisConcatenator):
+
+    def __init__(self):
+        AxisConcatenator.__init__(self, -1, ndmin=2, trans1d=0)
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: How about using `super`?

File: cupy/indexing/generate.py [hunk 4]
```diff
@@ -1,13 +1,167 @@
...
+
+c_ = CClass()
+"""Translates slice objects to concatenation along the second axis.
+
+This is a CuPy object that corresponds to :func:`cupy.r_`, which is
+useful because of its common occurrence. In particular, arrays will be
+stacked along their last axis after being upgraded to at least 2-D with
+1's post-pended to the shape (column vectors made out of 1-D arrays).
+
+For detailed documentation, see `r_`.
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: :func:`cupy.r_`?

File: cupy/indexing/generate.py [hunk 5]
```diff
@@ -1,13 +1,167 @@
...
+
+This implementation is partially borrowed from NumPy's one
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Append a period.

File: cupy/indexing/generate.py [hunk 6]
```diff
@@ -1,13 +1,167 @@
...
+>>> cupy.c_[a, 0, 0, b]
+array([[1, 2, 3, 0, 0, 4, 5, 6]])
+
+"""
+
+
+class RClass(AxisConcatenator):
+
+    def __init__(self):
+        AxisConcatenator.__init__(self)
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: How about using `super`?

File: cupy/indexing/generate.py [hunk 7]
```diff
@@ -1,13 +1,167 @@
...
+If the index expression contains comma separated arrays, then stack
+them along their first axis.
+
+This object can build up from normal CuPy arrays.
+Therefore, the other objects (e.g. writing strings like '2,3,4',
+or using imaginary numbers like [1,2,3j],
+or using string integers like '-1') are not implemented yet
+compared with NumPy.
+
+This implementation is partially borrowed from NumPy's one
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Append a period.

File: cupy/indexing/generate.py [hunk 8]
```diff
@@ -1,13 +1,167 @@
...
+
 import cupy
+
+import six
+
+
+class AxisConcatenator(object):
+    """Translates slice objects to concatenation along an axis.
+
+    For detailed documentation on usage, see `r_`.
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Use :func:

File: cupy/indexing/generate.py [hunk 9]
```diff
@@ -1,13 +1,167 @@
...
+    This implementation is partially borrowed from NumPy's one
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Append a period.

File: cupy/indexing/generate.py [hunk 10]
```diff
@@ -1,13 +1,167 @@
...
+                raise NotImplementedError
+            elif type(key[k]) in numpy.ScalarType:
+                newobj = cupy.array(key[k], ndmin=ndmin)
+                scalars.append(k)
+                scalar = True
+                scalartypes.append(newobj.dtype)
+            else:
+                newobj = key[k]
+                tempobj = cupy.array(newobj, copy=False)
+                newobj = cupy.array(newobj, copy=False, ndmin=ndmin)
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Why do you use the same name?

File: cupy/indexing/generate.py [hunk 11]
```diff
@@ -1,13 +1,167 @@
...
+                        'special directives must be the first entry.')
+                raise NotImplementedError
+            elif type(key[k]) in numpy.ScalarType:
+                newobj = cupy.array(key[k], ndmin=ndmin)
+                scalars.append(k)
+                scalar = True
+                scalartypes.append(newobj.dtype)
+            else:
+                newobj = key[k]
+                tempobj = cupy.array(newobj, copy=False)
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Is this `tempobj` required?

File: cupy/indexing/generate.py [hunk 12]
```diff
@@ -1,13 +1,167 @@
...
+                newobj = cupy.array(newobj, copy=False, ndmin=ndmin)
+                if ndmin > 1:
+                    if trans1d != -1 and tempobj.ndim < ndmin:
+                        newobj = self._output_obj(newobj, ndmin, trans1d)
+                    del tempobj
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Why don't you always delete `tempobj` ?

File: cupy/indexing/generate.py [hunk 13]
```diff
@@ -1,13 +1,167 @@
...
 # class s_(object):
 
 import numpy
+
 import cupy
+
+import six
+
+
+class AxisConcatenator(object):
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Is this class public? Are default arguments for its initializer required?

File: cupy/indexing/generate.py (hunk 11 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: I refactored the code.

File: cupy/indexing/generate.py (hunk 3 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: Fixed

File: cupy/indexing/generate.py (hunk 4 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: Fixed

File: cupy/indexing/generate.py (hunk 5 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: Fixed

File: cupy/indexing/generate.py (hunk 6 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: Fixed

File: cupy/indexing/generate.py (hunk 7 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: Fixed

File: cupy/indexing/generate.py (hunk 8 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: Fixed

File: cupy/indexing/generate.py (hunk 2 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: Fixed

File: cupy/indexing/generate.py (hunk 10 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: I refactored the code.

**Comment by delta2323** [comment]
Thank you for your comment. I fixed the code.

File: cupy/indexing/generate.py [hunk 14]
```diff
@@ -1,13 +1,159 @@
...
-
-# class c_(object):
-# class r_(object):
 # class s_(object):
 
 import numpy
+
 import cupy
+
+import six
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Fix import orde

File: cupy/indexing/generate.py [hunk 15]
```diff
@@ -1,13 +1,159 @@
...
+
+
+class AxisConcatenator(object):
+    """Translates slice objects to concatenation along an axis.
+
+    For detailed documentation on usage, see :func:`r_`.
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: `:func:cupy.r_`?

File: cupy/indexing/generate.py (hunk 13 above)
Comment (unnonouno) [inline cupy/indexing/generate.py]: CClass specifies all arguments and RClass uses the default values. This implementation looks strange.

File: cupy/indexing/generate.py [hunk 16]
```diff
@@ -1,13 +1,159 @@
...
+
+        final_dtype = numpy.find_common_type(arraytypes, scalartypes)
+        if final_dtype is not None:
+            for k in scalars:
+                objs[k] = objs[k].astype(final_dtype)
+
+        return cupy.concatenate(tuple(objs), axis=self.axis)
+
+    def __len__(self):
+        return 0
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Please test this method.

File: cupy/indexing/generate.py (hunk 13 above)
Comment (delta2323) [inline cupy/indexing/generate.py]: In NumPy, `RClass` does call `AxisConcatenator.__init__(self, 0)` (see [here](https://github.com/numpy/numpy/blob/master/numpy/lib/index_tricks.py#L454)). But as `0` is same as the default argument, it does not have any effect. The current PR is the result of the removal of this argument.

But as you point out, `AxisConcatenator` and `RClass` are equivalent. If we do not mind the inconsistency about the class hierarchies between NumPy and CuPy and non-existence of `AxisConcatenator`, it is OK to remove `AxisConcatenator` and move its implementation to `RClass`.

**Comment by unnonouno** [comment]
Can you check flake8 and merge master branch?

File: cupy/indexing/generate.py [hunk 17]
```diff
@@ -1,13 +1,163 @@
...
+        k2 = ndmin - ndim
+        if trans1d < 0:
+            trans1d += k2 + 1
+        defaxes = list(six.moves.range(ndmin))
+        k1 = trans1d
+        axes = defaxes[:k1] + defaxes[k2:] + \
+            defaxes[k1:k2]
+        return obj.transpose(axes)
+
+    def __init__(self, axis=0, matrix=False, ndmin=1, trans1d=-1):
```
Comment (unnonouno) [inline cupy/indexing/generate.py]: Are these default arguments really required? They are overwritten by its subclass. Users may use this class directly?

Comment (delta2323) [inline cupy/indexing/generate.py]: It is better to set default arguments to align the API with that of NumPy. But I do not think `AxisConcatenator` is expected to be used by users directly in NumPy  is not in `__all__` (see [here](https://github.com/numpy/numpy/blob/master/numpy/lib/index_tricks.py#L21)), although it is not prefixed with underscores.

Although it is documented and not prefixed with underscores, .

Comment (unnonouno) [inline cupy/indexing/generate.py]: OK

**Comment by unnonouno** [comment]
OK, LGTM

**Comment by delta2323** [comment]
Thank you.

**Pull request merged**